README.md
[![R-CMD-check](https://github.com/ropensci/git2r/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/git2r/actions)
[![CRAN status](http://www.r-pkg.org/badges/version/git2r)](http://cran.r-project.org/web/packages/git2r/index.html)
[![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/last-month/git2r)](http://cran.r-project.org/web/packages/git2r/index.html)
[![Coverage Status](https://coveralls.io/repos/github/ropensci/git2r/badge.svg?branch=master)](https://coveralls.io/github/ropensci/git2r?branch=master)

# Introduction

The `git2r` package gives you programmatic access to Git repositories
from R. Internally the package uses the libgit2 library which is a
pure C implementation of the Git core methods. For more information
about libgit2, check out libgit2's website
[(http://libgit2.github.com)](http://libgit2.github.com).

Suggestions, bugs, forks and pull requests are appreciated. Get in
touch.

## Installation

To install the version available on CRAN:

```coffee
install.packages("git2r")
```

To install the development version of `git2r`, it's easiest to use the
devtools package:

```coffee
# install.packages("devtools")
library(devtools)
install_github("ropensci/git2r")
```

Another alternative is to use `git` and `make`

```coffee
$ git clone https://github.com/ropensci/git2r.git
$ cd git2r
$ make install
```

## Usage

### Repository

The central object in the `git2r` package is the S3 class
`git_repository`. The following three methods can instantiate a
repository; `init`, `repository` and `clone`.

#### Create a new repository

Create a new repository in a temporary directory using `init`

```coffee
library(git2r)
```

```
#> Loading required package: methods
```

```coffee

## Create a temporary directory to hold the repository
path <- tempfile(pattern="git2r-")
dir.create(path)

## Initialize the repository
repo <- init(path)

## Display a brief summary of the new repository
repo
```

```
#> Local:    /tmp/Rtmp7CXPlx/git2r-1ae2305c0e8d/
#> Head:     nothing commited (yet)
```

```coffee

## Check if repository is bare
is_bare(repo)
```

```
#> [1] FALSE
```

```coffee

## Check if repository is empty
is_empty(repo)
```

```
#> [1] TRUE
```

#### Create a new bare repository

```coffee
## Create a temporary directory to hold the repository
path <- tempfile(pattern="git2r-")
dir.create(path)

## Initialize the repository
repo <- init(path, bare=TRUE)

## Check if repository is bare
is_bare(repo)
```

```
#> [1] TRUE
```

#### Clone a repository

```coffee
## Create a temporary directory to hold the repository
path <- file.path(tempfile(pattern="git2r-"), "git2r")
dir.create(path, recursive=TRUE)

## Clone the git2r repository
repo <- clone("https://github.com/ropensci/git2r", path)
```

```
#> cloning into '/tmp/Rtmp7CXPlx/git2r-1ae27d811539/git2r'...
#> Receiving objects:   1% (24/2329),   12 kb
#> Receiving objects:  11% (257/2329),   60 kb
#> Receiving objects:  21% (490/2329),  100 kb
#> Receiving objects:  31% (722/2329),  125 kb
#> Receiving objects:  41% (955/2329),  237 kb
#> Receiving objects:  51% (1188/2329),  574 kb
#> Receiving objects:  61% (1421/2329), 1014 kb
#> Receiving objects:  71% (1654/2329), 1350 kb
#> Receiving objects:  81% (1887/2329), 1733 kb
#> Receiving objects:  91% (2120/2329), 2614 kb
#> Receiving objects: 100% (2329/2329), 2641 kb, done.
```

```coffee

## Summary of repository
summary(repo)
```

```
#> Remote:   @ origin (https://github.com/ropensci/git2r)
#> Local:    master /tmp/Rtmp7CXPlx/git2r-1ae27d811539/git2r/
#>
#> Branches:          1
#> Tags:              0
#> Commits:         320
#> Contributors:      3
#> Ignored files:     0
#> Untracked files:   0
#> Unstaged files:    0
#> Staged files:      0
```

```coffee

## List all references in repository
references(repo)
```

```
#> $`refs/heads/master`
#> [6fb440] master
#>
#> $`refs/remotes/origin/master`
#> [6fb440] origin/master
```

```coffee

## List all branches in repository
branches(repo)
```

```
#> [[1]]
#> [6fb440] (Local) (HEAD) master
#>
#> [[2]]
#> [6fb440] (origin @ https://github.com/ropensci/git2r) master
```

#### Open an existing repository

```coffee
## Open an existing repository
repo <- repository(path)

## Workdir of repository
workdir(repo)
```

```
#> [1] "/tmp/Rtmp7CXPlx/git2r-1ae27d811539/git2r/"
```

```coffee

## List all commits in repository
commits(repo)[[1]] # Truncated here for readability
```

```
#> Commit:  6fb440133765e80649de8d714eaea17b114bd0a7
#> Author:  Stefan Widgren <stefan.widgren@gmail.com>
#> When:    2014-04-22 21:43:19
#> Summary: Fixed clone progress to end line with newline
```

```coffee

## Get HEAD of repository
repository_head(repo)
```

```
#> [6fb440] (Local) (HEAD) master
```

```coffee

## Check if HEAD is head
is_head(repository_head(repo))
```

```
#> [1] TRUE
```

```coffee

## Check if HEAD is local
is_local(repository_head(repo))
```

```
#> [1] TRUE
```

```coffee

## List all tags in repository
tags(repo)
```

```
#> list()
```

### Configuration

```coffee
config(repo, user.name="Git2r Readme", user.email="git2r.readme@example.org")

## Display configuration
config(repo)
```

```
#> global:
#>         core.autocrlf=input
#> local:
#>         branch.master.merge=refs/heads/master
#>         branch.master.remote=origin
#>         core.bare=false
#>         core.filemode=true
#>         core.logallrefupdates=true
#>         core.repositoryformatversion=0
#>         remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
#>         remote.origin.url=https://github.com/ropensci/git2r
#>         user.email=git2r.readme@example.org
#>         user.name=Git2r Readme
```

### Commit

```coffee
## Create a new file
writeLines("Hello world!", file.path(path, "test.txt"))

## Add file and commit
add(repo, "test.txt")
commit(repo, "Commit message")
```

```
#> Commit:  0a6af48cedf43208bde34230662280514e0956eb
#> Author:  Git2r Readme <git2r.readme@example.org>
#> When:    2014-04-22 21:44:57
#> Summary: Commit message
```

# Included software

- The C library [libgit2](https://github.com/libgit2/libgit2). See
  `inst/AUTHORS` for the authors of libgit2.

- The libgit2 library has been modified, e.g. to use the R printing
  and error routines, and to use `runif` instead of `rand`.

# License

The `git2r` package is licensed under the GPLv2. See these files for additional details:

- LICENSE      - `git2r` package license (GPLv2)
- inst/COPYING - Copyright notices for additional included software

---

[![](http://ropensci.org/public_images/github_footer.png)](http://ropensci.org)


NEWS
git2r (development version)
---------------------------

git2r 0.28.0
------------

IMPROVEMENTS

* Updated to use libgit2 version 1.1.0 on Windows.

* Fix handling of a symbolic reference when checking out previous
  branch.

* Added a configure option '--without-libgit2' to ignore presence of a
  system libgit2 library and instead use the internal git2r libgit2
  library. Usage:
  R CMD INSTALL --configure-args='--without-libgit2' git2r_x.y.z.tar.gz

* Updated some tests to work with libgit2 version 1.1.0.

git2r 0.27.1
------------

CHANGES

* Fixed the CITATION file to pass 'R CMD check' without a NOTE.

git2r 0.27.0
------------

IMPROVEMENTS

* Updated the bundled libgit2 source code to version '0.28.5'.

* Updated the build configuration script to be able to build git2r
  with a system installation of libgit2 version >= 1.0.

* Updated to use libgit2 version 1.0.0 on Windows.

* The build configuration script checks for minimum required version
  of libssh2 (version >= 1.8). Issue #420.

* Updated to use roxygen2 version 7.1.0 to build the documentation.

* Make it easier to view and change the timezone (John Blischak in
  #408).

* Fixed 'ls_tree' to handle content in subfolder, see description in
  PR #402.

* The 'branch_create' function has been changed to use the
  'last_commit()' function as default to determine the commit to which
  the new branch should point.

git2r 0.26.1
------------

BUG FIXES

* Fixed the broken build on Solaris.

git2r 0.26.0
------------

IMPROVEMENTS

* Updated the bundled libgit2 source code to version '0.28.2'.

* Added the 'force' argument to the 'tag' function to overwrite an
  existing tag.

* Allow a zero length tag message.

* Make it possible to create a lighweight tag.

* Added the 'ref' argument to the 'commits' function to give a
  reference to list commits from.

* Added the utility function 'lookup_commit' to lookup a commit
  related to a git object.

* The 'path' argument was added to the 'commits' function to make it
  possible to specify that only commits modifying this file ('path')
  will be returned to reproduce 'git log' with '--no-follow', see the
  documentation. (Peter Carbonetto and John Blischak in PR #372)

BUG FIXES

* Removed the timezone offset from the commit time to fix an incorrect
  time in GMT when reading information from a repository (Thierry
  Onkelinx in PR #393).

git2r 0.25.2
------------

CHANGES

* Improved the build configuration script: if the system installation
  of libgit2 is to old, use the bundled libgit2 instead of raising an
  error.

BUG FIXES

* Fixed the broken build on Solaris.

git2r 0.25.1
------------

BUG FIXES

* Fixed significant warning from 'R CMD check'

git2r 0.25.0
------------

CHANGES

* Updated the bundled libgit2 source code to version '0.28.1'.

* Added additional parameters to the 'diff' function to control the
  output, see the documentation.

* Added getPass option to the password argument in 'cred_user_pass'
  (Annie Wang in PR #383)

* Changed the 'print' functions to return its argument invisibly.

* Changed the 'git_config_files' function to return a 'data.frame'

* Changed the 'ahead_behind' function to accept a tag or a branch for
  the local and upstrean commit.

* Changed the 'descendent_of' function to accept a tag or a branch for
  the 'commit' and 'ancestor' commit.

BUG FIXES

* Fixed memory protection errors in the git2r C source code reported
  by the 'rchk' tool.

* Fixed listing of 'commits' from a shallow repository.

* Fixed the configuration script to include the missing macro
  'AM_ICONV'.

git2r 0.24.0
------------

This is a bug-fix release.

BUG FIXES

* Fixed memory protection errors in the git2r C source code reported
  by the 'rchk' tool.

* Raise an error if the path argument to the 'hashfile' function is
  NA.

git2r 0.23.0
------------

IMPROVEMENTS

* Updated the bundled libgit2 source code to v0.27.3 (504bd54).

BREAKING CHANGE

* On macOS, git2r no longer enables SSH transport by default. This is
  due to the complexity to build the dependencies for SSH transport in
  an R package when macOS no longer ships the OpenSSL headers.
  However, you can install git2r from source on macOS (see the 'R
  Installation and Administration' manual) with SSH transport enabled
  if you first install the libgit2 library, for example, using the
  Homebrew package manager. Another possibility is to let the build
  configuration automatically download the libgit2 library from the
  Homebrew package manager with:

  install.packages('git2r', type='source', configure.vars='autobrew=yes')

git2r 0.22.1
------------

NEW FEATURES

* Added the 'git_config_files' method to locate configuration files.

* Added the 'stash_pop' method to apply a single stashed state from
  the stash list and remove it from the list if successful.

* Added the 'stash_apply' method to apply a single stashed state from
  the stash list.

IMPROVEMENTS

* Updated the bundled libgit2 source code to v0.27.2 (8d36dc6).

* git2r can now build against a system installation of libgit2
  (Elliott Sales de Andrade in PR #345, #344 and #336).

* Refactoring of the configuration scripts to use a prebuilt libgit2
  on macOS and Windows (Thanks Jeroen).

* Ensure that git2r writes the config file to the correct location on
  Windows (John Blischak in PR #320).

* Better default location to find ssh keys in 'cred_ssh_key()' (Ian
  Lyttle in PR #317).

BUG FIXES

* If a merge results in no change, the returned 'git_merge_result' now returns
  'FALSE' for 'fast_forward' and 'conflicts' and 'NA' for 'sha'. Previously it
  returned 'logical(0)' for 'fast_forward' and 'conflicts' and 'character(0)'
  for 'sha'.

BREAKING CHANGES

* Changed from S4 classes to S3 classes to simplify the design and
  facilitate future development.

* Removed the trailing slash from the directory name when reporting
  repository path or workdir.

* Removed the 'libgit2_sha' method. Use the 'libgit2_version' method
  instead.


R/diff.r
## git2r, R bindings to the libgit2 library.
## Copyright (C) 2013-2015 The git2r contributors
##
## This program is free software; you can redistribute it and/or modify
## it under the terms of the GNU General Public License, version 2,
## as published by the Free Software Foundation.
##
## git2r is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License along
## with this program; if not, write to the Free Software Foundation, Inc.,
## 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

#' Number of files in git_diff object
#'
#' @docType methods
#' @param x The git_diff \code{object}
#' @return a non-negative integer
#' @keywords methods
#' @export
setMethod("length",
          signature(x = "git_diff"),
          function(x)
          {
              length(x@files)
          }
)

##' Show a diff
##'
##' @aliases show,git_diff-methods
##' @docType methods
##' @param object The diff \code{object}.
##' @keywords methods
##' @export
##' @examples
##' \dontrun{
##' ## Initialize a repository
##' path <- tempfile(pattern="git2r-")
##' dir.create(path)
##' repo <- init(path)
##'
##' ## Config user
##' config(repo, user.name="Alice", user.email="alice@@example.org")
##'
##' ## Create a file, add, commit
##' writeLines("Hello world!", file.path(path, "test.txt"))
##' add(repo, "test.txt")
##' commit(repo, "Commit message")
##'
##' ## Change the file
##' writeLines(c("Hello again!", "Here is a second line", "And a third"),
##'            file.path(path, "test.txt"))
##'
##' ## Brief summary of diff between index and workdir
##' diff(repo)
##' }
setMethod("show",
          signature(object = "git_diff"),
          function(object)
          {
              cat("Old:  ")
              if (is.character(object@old)) {
                  cat(object@old, "\n")
              } else if (is(object@old, "git_tree")) {
                  show(object@old)
              } else {
                  cat("\n")
                  show(object@old)
              }
              cat("New:  ")
              if (is.character(object@new)) {
                  cat(object@new, "\n")
              } else if (is(object@new, "git_tree")) {
                  show(object@new)
              } else {
                  cat("\n")
                  show(object@new)
              }
          }
)

lines_per_file <- function(diff) {
    lapply(diff@files, function(x) {
        del <- add <- 0
        for (h in x@hunks) {
            for (l in h@lines) {
                if (l@origin == 45) {
                    del <- del + l@num_lines
                } else if (l@origin == 43) {
                    add <- add + l@num_lines
                }
            }
        }
        list(file=x@new_file, del=del, add=add)
    })
}

print_lines_per_file <- function(diff) {
  lpf <- lines_per_file(diff)
  files <- vapply(lpf, function(x) x$file, character(1))
  del <- vapply(lpf, function(x) x$del, numeric(1))
  add <- vapply(lpf, function(x) x$add, numeric(1))
  paste0(format(files), " | ", "-", format(del), " +", format(add))
}

hunks_per_file <- function(diff) {
    vapply(diff@files, function(x) length(x@hunks), numeric(1))
}

##' Show the summary of a diff
##'
##' @aliases summary,git_diff-methods
##' @docType methods
##' @param object The diff \code{object}.
##' @param ... Additional arguments affecting the summary produced.
##' @return None (invisible 'NULL').
##' @keywords methods
##' @export
##' @examples
##' \dontrun{
##' ## Initialize a repository
##' path <- tempfile(pattern="git2r-")
##' dir.create(path)
##' repo <- init(path)
##'
##' ## Config user
##' config(repo, user.name="Alice", user.email="alice@@example.org")
##'
##' ## Create a file, add, commit
##' writeLines("Hello world!", file.path(path, "test.txt"))
##' add(repo, "test.txt")
##' commit(repo, "Commit message")
##'
##' ## Change the file
##' writeLines(c("Hello again!", "Here is a second line", "And a third"),
##'            file.path(path, "test.txt"))
##'
##' ## diff between index and workdir
##' diff_1 <- diff(repo)
##' summary(diff_1)
##'
##' ## Diff between index and HEAD is empty
##' diff_2 <- diff(repo, index=TRUE)
##' summary(diff_2)
##'
##' ## Diff between tree and working dir, same as diff_1
##' diff_3 <- diff(tree(commits(repo)[[1]]))
##' summary(diff_3)
##'
##' ## Add changes, diff between index and HEAD is the same as diff_1
##' add(repo, "test.txt")
##' diff_4 <- diff(repo, index=TRUE)
##' summary(diff_4)
##'
##' ## Diff between tree and index
##' diff_5 <- diff(tree(commits(repo)[[1]]), index=TRUE)
##' summary(diff_5)
##'
##' ## Diff between two trees
##' commit(repo, "Second commit")
##' tree_1 <- tree(commits(repo)[[2]])
##' tree_2 <- tree(commits(repo)[[1]])
##' diff_6 <- diff(tree_1, tree_2)
##' summary(diff_6)
##'
##' ## Binary files
##' set.seed(42)
##' writeBin(as.raw((sample(0:255, 1000, replace=TRUE))),
##'          con=file.path(path, "test.bin"))
##' add(repo, "test.bin")
##' diff_7 <- diff(repo, index=TRUE)
##' summary(diff_7)
##' }
setMethod("summary",
          signature(object = "git_diff"),
          function(object, ...)
          {
              show(object)
              if (length(object) > 0) {
                  plpf <- print_lines_per_file(object)
                  hpf <- hunks_per_file(object)
                  hunk_txt <- ifelse(hpf > 1, " hunks",
                                     ifelse(hpf > 0, " hunk",
                                            " hunk (binary file)"))
                  phpf <- paste0("  in ", format(hpf), hunk_txt)
                  cat("Summary:", paste0(plpf, phpf), sep="\n")
              } else {
                  cat("No changes.\n")
              }
          }
)

#' Changes between commits, trees, working tree, etc.
#'
#' @rdname diff-methods
#' @export
#' @docType methods
#' @param x A \code{git_repository} object or the old
#' \code{git_tree} object to compare to.
#' @param index \describe{
#'   \item{\emph{When object equals a git_repository}}{
#'     Whether to compare the index to HEAD. If FALSE (the default),
#'     then the working tree is compared to the index.
#'   }
#'   \item{\emph{When object equals a git_tree}}{
#'     Whether to use the working directory (by default), or the index
#'     (if set to TRUE) in the comparison to \code{object}.
#'   }
#' }
#' @param as_char logical: should the result be converted to a
#' character string?. Default is FALSE.
#' @param filename If as_char is TRUE, then the diff can be written to
#' a file with name filename (the file is overwritten if it
#' exists). Default is NULL.
#' @param ... Additional arguments affecting the diff produced
#' @return A \code{\linkS4class{git_diff}} object if as_char is
#' FALSE. If as_char is TRUE and filename is NULL, a character string,
#' else NULL.
#' @keywords methods
#' @examples
#' \dontrun{
#' ## Initialize a repository
#' path <- tempfile(pattern="git2r-")
#' dir.create(path)
#' repo <- init(path)
#'
#' ## Config user
#' config(repo, user.name="Alice", user.email="alice@@example.org")
#'
#' ## Create a file, add, commit
#' writeLines("Hello world!", file.path(path, "test.txt"))
#' add(repo, "test.txt")
#' commit(repo, "Commit message")
#'
#' ## Change the file
#' writeLines(c("Hello again!", "Here is a second line", "And a third"),
#'            file.path(path, "test.txt"))
#'
#' ## diff between index and workdir
#' diff_1 <- diff(repo)
#' summary(diff_1)
#' cat(diff(repo, as_char=TRUE))
#'
#' ## Diff between index and HEAD is empty
#' diff_2 <- diff(repo, index=TRUE)
#' summary(diff_2)
#' cat(diff(repo, index=TRUE, as_char=TRUE))
#'
#' ## Diff between tree and working dir, same as diff_1
#' diff_3 <- diff(tree(commits(repo)[[1]]))
#' summary(diff_3)
#' cat(diff(tree(commits(repo)[[1]]), as_char=TRUE))
#'
#' ## Add changes, diff between index and HEAD is the same as diff_1
#' add(repo, "test.txt")
#' diff_4 <- diff(repo, index=TRUE)
#' summary(diff_4)
#' cat(diff(repo, index=TRUE, as_char=TRUE))
#'
#' ## Diff between tree and index
#' diff_5 <- diff(tree(commits(repo)[[1]]), index=TRUE)
#' summary(diff_5)
#' cat(diff(tree(commits(repo)[[1]]), index=TRUE, as_char=TRUE))
#'
#' ## Diff between two trees
#' commit(repo, "Second commit")
#' tree_1 <- tree(commits(repo)[[2]])
#' tree_2 <- tree(commits(repo)[[1]])
#' diff_6 <- diff(tree_1, tree_2)
#' summary(diff_6)
#' cat(diff(tree_1, tree_2, as_char=TRUE))
#'
#' ## Binary files
#' set.seed(42)
#' writeBin(as.raw((sample(0:255, 1000, replace=TRUE))),
#'          con=file.path(path, "test.bin"))
#' add(repo, "test.bin")
#' diff_7 <- diff(repo, index=TRUE)
#' summary(diff_7)
#' cat(diff(repo, index=TRUE, as_char=TRUE))
#' }
setMethod("diff",
          signature(x = "git_repository"),
          function(x,
                   index    = FALSE,
                   as_char  = FALSE,
                   filename = NULL,
                   ...)
          {
              if (as_char) {
                  ## Make sure filename is character(0) to write to a
                  ## character vector or a character vector with path in
                  ## order to write to a file.
                  filename <- as.character(filename)
                  if (any(identical(filename, NA_character_),
                          identical(nchar(filename), 0L))) {
                      filename <- character(0)
                  } else if (length(filename)) {
                      filename <- normalizePath(filename, mustWork = FALSE)
                  }
              } else {
                  ## Make sure filename is NULL
                  filename <- NULL
              }

              .Call(git2r_diff, x, NULL, NULL, index, filename)
          }
)

#' @rdname diff-methods
#' @param new_tree The new git_tree object to compare, or NULL.
#'        If NULL, then we use the working directory or the index (see
#'        the \code{index} argument).
#' @export
setMethod("diff",
          signature(x = "git_tree"),
          function(x,
                   new_tree = NULL,
                   index    = FALSE,
                   as_char  = FALSE,
                   filename = NULL,
                   ...)
          {
              if (as_char) {
                  ## Make sure filename is character(0) to write to a
                  ## character vector or a character vector with path in
                  ## order to write to a file.
                  filename <- as.character(filename)
                  if (any(identical(filename, NA_character_),
                          identical(nchar(filename), 0L))) {
                      filename <- character(0)
                  } else if (length(filename)) {
                      filename <- normalizePath(filename, mustWork = FALSE)
                  }
              } else {
                  ## Make sure filename is NULL
                  filename <- NULL
              }

              if (!is.null(new_tree)) {
                  if (! is(new_tree, "git_tree")) {
                      stop("Not a git tree")
                  }
                  if (x@repo@path != new_tree@repo@path) {
                      stop("Cannot compare trees in different repositories")
                  }
              }

              .Call(git2r_diff, NULL, x, new_tree, index, filename)
          }
)
